### PR TITLE
[TTS] Add config and modules for 22khz and 44khz audio codec

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Add config files and corresponding modules for optimized audio codec training.

**Collection**: [TTS]

# Changelog
- Add config files for training 22khz and 44khz audio codec.
- Add inverted HiFiGAN audio encoder
- Add multi resolution STFT discriminator from DAC. This does not make much difference by itself, but I changed in preparation for doing mixed bandwidth training.
- Added a "half snake" activation to decoder which uses half snake and half leaky relu activations, allowing the model to jointly model both periodic and non-periodic information. When tuning hyperparameters, I found this to be more stable (and use less memory) than training with all snake activations. And perform better than using the existing _snake alpha_, and had similar performance to the _snake beta_ variants from BigVGAN 1 and 2. I did not add snake activation to encoder as doing so significantly reduced TTS performance on the codec.
- Replaced tanh activation on output with clamping, as in BigVGAN 2, as this seemed to reduce artifacts early in training.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

**Review by Edresson — DISMISSED** [review]

File: examples/tts/conf/audio_codec/audio_codec_22050.yaml [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# This config contains the default values for training 44.1kHz audio codec model which encodes mel spectrogram
+# instead of raw audio.
+# If you want to train model on other dataset, you can change config values according to your dataset.
+# Most dataset-specific arguments are in the head of the config file, see below.
+
+name: AudioCodec
```
Comment (nithinraok) [inline examples/tts/conf/audio_codec/audio_codec_22050.yaml:6]: Should this file be part of audio collection or tts collection? CC: @anteju

File: examples/tts/conf/audio_codec/audio_codec_22050.yaml [hunk 2]
```diff
@@ -0,0 +1,194 @@
+# This config contains the default values for training 44.1kHz audio codec model which encodes mel spectrogram
```
Comment (nithinraok) [inline examples/tts/conf/audio_codec/audio_codec_22050.yaml:1]:
```suggestion
# This config contains the default values for training 22.5kHz audio codec model which encodes mel spectrogram
```

File: examples/tts/conf/audio_codec/audio_codec_44100.yaml [hunk 3]
```diff
@@ -0,0 +1,194 @@
+# This config contains the default values for training 44.1kHz audio codec model which encodes mel spectrogram
+# instead of raw audio.
```
Comment (nithinraok) [inline examples/tts/conf/audio_codec/audio_codec_44100.yaml:2]: Is this mel spectrogram codec or raw audio codec?

File: nemo/collections/asr/parts/utils/activations.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
 import torch
 import torch.nn as nn
 
-__all__ = ['Swish', 'Snake']
+__all__ = ['Swish', 'Snake', 'HalfSnake']
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/activations.py:18]: Wondering if we should move Snake and HalfSnake to audio collection instead of asr

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 5]
```diff
@@ -322,6 +324,152 @@ def forward(self, audio_real, audio_gen):
         return scores_real, scores_gen, fmaps_real, fmaps_gen
 
 
+class DiscriminatorSTFT(NeuralModule):
```
Comment (nithinraok) [inline nemo/collections/tts/modules/audio_codec_modules.py:327]: Docs missing

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 6]
```diff
@@ -322,6 +324,152 @@ def forward(self, audio_real, audio_gen):
...
+            fmap.append(out)
+        # [batch, 1, T_spec, fft // 8]
+        scores = self.conv_post(inputs=out)
+        fmap.append(scores)
+        scores = rearrange(scores, "B 1 T C -> B C T")
+
+        return scores, fmap
+
+
+class MultiBandDiscriminatorSTFT(NeuralModule):
```
Comment (nithinraok) [inline nemo/collections/tts/modules/audio_codec_modules.py:375]: Docs missing

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: examples/tts/conf/audio_codec/audio_codec_22050.yaml [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    lr: 2e-4
+    betas: [0.8, 0.99]
+
+    sched:
+      name: ExponentialLR
+      gamma: 0.998
+
+trainer:
+  num_nodes: 1
+  devices: 1
```
[the remark this answers is not in the document]
Comment (nithinraok) [reply examples/tts/conf/audio_codec/audio_codec_22050.yaml:166]: I would recommend to put the same number of devices and optimizer values etc with which the public model was trained unless we optimized a model for 1 GPU

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 8]
```diff
@@ -933,7 +1190,12 @@ def __init__(
 
         self.post_activation = CodecActivation(activation, channels=in_channels)
         self.post_conv = Conv1dNorm(in_channels=in_channels, out_channels=1, kernel_size=out_kernel_size)
-        self.out_activation = nn.Tanh()
+        if output_activation == "tanh":
+            self.out_activation = nn.Tanh()
+        elif output_activation == "clamp":
+            self.out_activation = AudioClampActivation()
+        else:
+            raise ValueError(f"Invalid audio output activation {output_activation}")
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:1198]: Is using the registry a good idea if other activation functions would not provide a valid audio signal as output?

File: nemo/collections/asr/parts/utils/activations.py (hunk 4 above)
Comment (rlangman) [reply nemo/collections/asr/parts/utils/activations.py:18]: If I add it to common, should it be in the `activation_registry`? The registry only seems useful if the activation does not require input arguments.

Otherwise, it seems roundabout to have something like:

```
if activation in ["snake", "half_snake"]:
  self.activation = activation_registry[activation](channels)
else:
  self.activation = activation_registry[activation]()
```

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 5 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:326]: Added

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 6 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:378]: Added

File: examples/tts/conf/audio_codec/audio_codec_22050.yaml (hunk 2 above)
Comment (rlangman) [reply examples/tts/conf/audio_codec/audio_codec_22050.yaml]: Fixed

File: examples/tts/conf/audio_codec/audio_codec_44100.yaml (hunk 3 above)
Comment (rlangman) [reply examples/tts/conf/audio_codec/audio_codec_44100.yaml]: Changed description to raw audio codec.

File: examples/tts/conf/audio_codec/audio_codec_22050.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (rlangman) [reply examples/tts/conf/audio_codec/audio_codec_22050.yaml]: I think it makes more sense to default to -1, unless we have an optimizer that really needs to be tuned based on batch size.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 9]
```diff
@@ -322,6 +324,152 @@ def forward(self, audio_real, audio_gen):
...
+        }
+
+    @typecheck()
+    def forward(self, spec):
+        fmap = []
+
+        # [batch, 2, T_spec, fft]
+        out = spec
+        for conv in self.conv_layers:
+            # [batch, filters, T_spec, fft // 2**i]
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py]: Changed to just say `fft // strides` as it looked to verbose to do something like `fft // mul(strides[:i+1])`

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 10]
```diff
@@ -868,6 +1016,114 @@ def forward(self, inputs, input_len):
...
+            # [B, 2 * C, T / down_sample_rate]
+            out = down_sample_conv(inputs=out, input_len=encoded_len)
+
+        out = self.post_activation(out)
+        # [B, encoded_dim, T_encoded]
+        encoded = self.post_conv(inputs=out, input_len=encoded_len)
+        return encoded, encoded_len
+
+
+class AudioClampActivation(nn.Module):
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py]: Changed

File: nemo/collections/tts/modules/audio_codec_modules.py (hunk 6 above)
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py:383]: Added

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 11]
```diff
@@ -868,6 +1028,108 @@ def forward(self, inputs, input_len):
         return out
 
 
+class HiFiGANEncoder(NeuralModule):
+    """
+    Encoder created by inverting the HiFi-GAN decoder
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/audio_codec_modules.py]: Added

**Review by Edresson — APPROVED** [review]

**Review by nithinraok — APPROVED** [review]
LGTM
